### ISSUE TITLE
Hertz application

### ISSUE THREAD

**Issue by MarijaI** [mfem/mfem#3221]
Hi MFEM friends,

Is it possible to modify Hertz application so that it can use port condition on one boundary face? Also, since I want to use absorbing conditions on two different materials, could I use `abcs1 `and `abcs2 `separately instead of using just one `abcs` array.
My idea was to add new conditions. Of course, I adjusted these changes in hertz.cpp, hertz_solver.hpp and herzt_solver.cpp and wrote corresponding functions in order to create coefficients.  

 Create the Magnetostatic solver
 ```
  HertzSolver Hertz(pmesh, order, freq_,
                     (HertzSolver::SolverType)sol, solOpts,
                     (HertzSolver::PrecondType)prec,
                     conv, *epsCoef, *muInvCoef, sigmaCoef, etaInvCoef1, etaInvCoef1,
                     abcs1, abcs2, dbcs, port,
                     e_bc_r, e_bc_i
                    );
```

Also, I have both RHS and LHS in my model. I've noticed that in Hertz there is no linear form on the LHS. 
I am solving the same problem as in https://github.com/mfem/mfem/issues/3069 only parallel version. 
This is error that I don't know how to overcome:
![image](https://user-images.githubusercontent.com/46717785/191197170-ff653bcb-4723-450a-bf14-bec93d9b9902.png)

Thanks in advance for any hint/help/idea
Marija

**Comment by MarijaI** [comment]
Hi again,

I found where the problem is. Code execution is interrupted when the command `RHS->Assemble()`  within 
`void HertzSolver::Assemble()`  function is run.

By the way, this is how RHS is defined. 
   ```
 RHS_ = new ParComplexLinearForm(HCurlFESpace_);
 RHS_->AddBoundaryIntegrator(new VectorFEDomainLFIntegrator(cEinc_Re),
        new VectorFEDomainLFIntegrator(cEinc_Im), port_marker_);  
 RHS_->AddBoundaryIntegrator(new VectorFEBoundaryTangentLFIntegrator(port_Re_Coef_),
        new VectorFEBoundaryTangentLFIntegrator(port_Im_Coef_), port_marker_);  
 RHS_->AddBoundaryIntegrator(NULL, new VectorFEBoundaryTangentLFIntegrator(cABS1), abc1_marker_);
 RHS_->AddBoundaryIntegrator(NULL, new VectorFEBoundaryTangentLFIntegrator(cABS1), abc2_marker_);
```
So, let me rephrase the question. Is it possible to modify the Hertz application to define the port conditions and to solve the system containing the ParSesquilinear form on the LHS and on ParComplexLinearForm on the RHS?

I assume that the problem arises when the coefficients are defined
```
double beta = 1;    // different value is in code
double muRc = 2;  // different value is in code
VectorFunctionCoefficient E_Re(pmesh_->SpaceDimension(), e_r_bc);
VectorFunctionCoefficient E_Im(pmesh_->SpaceDimension(), e_i_bc);

 ScalarVectorProductCoefficient cEinc_Re(-2. * beta/muRc, E_Im);
 ScalarVectorProductCoefficient cEinc_Im(2. * beta/muRc, E_Re);
```

Thanx again for any response.
Marija

**Comment by mlstowell** [comment]
Hello, @MarijaI,

It should be possible to assign a port boundary condition on any subset of boundary faces. I haven't had much experience with setting up port BCs in this code but it looks like you're doing this correctly.

Regarding absorbing BCs there are at least two ways to use different coefficients on different portions of the boundary. One is the way you've chosen which looks correct (although I notice you use `cABS1` on both portions which may be a typo). The other is to use a `Coefficient` object which returns different values on different portions of the boundary. Either method should be fine and the one you've selected is probably easier to setup.

Regarding the code failure I would suggest checking that the new `RHS_` object is being "updated" correctly. In my experience with AMR codes this is something I often overlook. Each AMR iteration has the potential to resize the mesh and consequently the various matrices and vectors need to be resized as well. This is done in the `HertzSolver::Update` method. There may be other possibilities but we should double-check this first.

Best wishes,
Mark

**Comment by MarijaI** [comment]
Hello @mlstowell ,
Thank you very much for your time and response.
Yes, I do have a typo in the code above. 
I update  `RHS_` object and the first two iteration goes well, but on the third iteration program crashes again.
The last information before program crashes is that `RHS_` is not Assembled in `HertzSolver::Assemble` (in the first two iteration everything is assembled fine).
I think that there is a problem with the coefficient. I will try to solve that by myself and ask for help in case I don't succeed. 
I just needed an info that it is possible to assign `ParComplexLinear `form on the left and that it is possible to work with it. 

Thank you one more time. Best,
Marija
